master/internal/api_workspace_test.go

```diff
@@ -0,0 +1,74 @@
+package internal
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestGenerateNamespaceName(t *testing.T) {
+	testCases := []struct {
+		name string
+
+		instNamespace string
+		clusterID     string
+		workspaceName string
+		workspaceID   uint
+
+		expectedName string
+	}{{
+		name:          "empty cluster id",
+		instNamespace: "foo",
+		clusterID:     "",
+		workspaceName: "bar",
+		workspaceID:   42,
+		expectedName:  "bar-foo--42",
+	}, {
+		name:          "empty workspace name",
+		instNamespace: "foo",
+		clusterID:     "bar",
+		workspaceName: "",
+		workspaceID:   742_915,
+		expectedName:  "-foo-bar-742915",
```

Comment:
This test case should actually fail because namespaces should start with an alpha-numeric character (ref [here](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names)). So if the workspace has no name, I think reasonably so, its auto-generated namespace creation should fail

Reply:
I guess if we don't want `generateNamespaceName` to error out at all, we could replace the empty string with the string`null`, which shouldn't cause issues with uniqueness given the clusterID and workspace ID appended to the name.

Reply:
Very interesting. Do you have an idea of how the function should behave when it can't begin with a workspace name? Perhaps just trim the `-` prefix? It currently can't return an error, but changing it to return one is also an option.

Reply:
I think if we change it to return an error, we can remove the cast to `uint` and assert within the function that the workspace id is a positive integer. Asserting valid parameters like the workspace name and ID could be useful for debugging from a backend perspective but, on the other hand, I like the idea of the auto-generation always working (kind of like a hash function). 
My only reluctance with removing the first hyphen is that the auto-created namespace might then look like its corresponding workspace name is the installation namespace name. 
The `workspace` table in the db has a non-empty name constraint [ref here](https://github.com/determined-ai/determined/blob/13aa327018f8b79d611f786568586fe385aceaa6/master/static/migrations/20220616002642_workspace-name-restrictions.tx.up.sql#L2), so I  kind of like handling the highly unlikely edge case of an empty name with the `null` string if we decide not the make the function error out.

Reply:
My apologies: I responded before seeing your second comment here. I agree!
- I'd like for this to always work like a hash without returning an error. Not only is that convenient at call time, but a function returning an `error` is often a signal that it does something which can fail like making a network or filesystem call. Since that's not the case here, it'd be a red herring.
- I'd rather check for arg validity with the compiler (i.e. accepting `uint` rather than returning an `error` at runtime)
- Using `null` in the string for clarity is a great idea.

I'll update this to use `null` in that slot when `workspaceName` is empty!

Reply:
Updated and also added a test for namespace name validity in the unit tests.

Reply:
Thanks! Agreed on all points